minpy/primitive.py

```diff
@@ -94,57 +128,62 @@ def get_val(x, mutate):
-            with current_context().as_mxnet_context() as ctx:
+            with context.current_context().as_mxnet_context():
                 result_value = self._func(*arg_values, **kwargs_values)
         else:
             result_value = self._func(*arg_values, **kwargs_values)
-        # if you want to do profiling, try to use minprof(<func>):
+        # If you want to do profiling, try to use `minprof(func)`.
         # result_value = minprof(self._func)(*arg_values, **kwargs_values)
 
-        # whether the result is on the bp path
         def scan(accum, x):
-            if isinstance(x, Value):
-                return operator.or_(accum, x._marked_for_bp)
+            """Whether the result is on the bp path."""
+            if isinstance(x, array.Value):
+                return operator.or_(accum, x.marked_for_bp)
             else:
                 return accum
-        # Check whether the result value is on the path of bp phase
+        # Check whether the result value is on the path of bp phase.
         # If all the input arguments are not on the bp path, the result value
         # is not as well.
         need_bp = functools.reduce(scan, itertools.chain(
             args, kwargs.values()), False)
         # Wrap the result raw value with wrapper and node.
-        result = Value.wrap(result_value, marked=need_bp)
+        result = array.Value.wrap(result_value, marked=need_bp)
         if need_bp:
-            # Record partial derivative paths, only for `Value` type values.
+            # Record partial derivative paths, only for `array.Value` type values.
             # If no gradient function is defined, also omit it
             for i, arg in enumerate(args):
-                if isinstance(arg, Value) and arg.marked_for_bp:
+                if isinstance(arg, array.Value) and arg.marked_for_bp:
                     if i not in self._grad_func:
-                        _logger.warn('Partial derivative of func "{}" on #{} \
-                            arg is not defined.'
-                                     .format(self._func.__name__, i))
+                        _logger.warning('Partial derivative of func "{}" on \
+                            argument {} is not defined.'
+                                        .format(self._func.__name__, i))
                         continue
                     _logger.debug(
-                        'Adding partial derivative to func "{}" on #{} arg.'.format(
-                            self._func, i))
-                    arg.node.add_partial_derivative(self._grad_func[i](
-                        result_value, *arg_values, **kwargs_values),
-                                                    result, self)
+                        'Adding partial derivative to func "{}" on argument {}.'.
+                        format(self._func, i))
+                    grad_func_rec = self._grad_func[i]
+                    grad_func = grad_func_rec.f(result_value, *arg_values,
```

Comment:
So the `f` field in `GradFunc` is like a gradient function selector? I would use another name.